### PR TITLE
Fix how item call numbers are derived

### PR DESCRIPTION
**Pull request by jthomale**
Changes call number derivation logic to fix problems identified through use of the inventory app. First, prioritizes actual call numbers (c-tagged fields in Sierra) over Sudoc/Gov Docs numbers (g-tagged fields) so that, if a call number and Sudoc number are present on the item or bib record, the call number will be used over the Sudoc number. Second, if a call number string exactly matches a Sudoc number string on the bib or item, the call number type is `sudoc`. Third, moves all call number derivation logic into a method on the ItemRecord model class instead of having it live in the ItemIndex class.

Fixes issue #11.

### CODE REVIEW

**Comment by jthomale** [comment]
@vphill @damonkelley @somexpert @jason-ellis @ldko  Again, feel free to take a look and review/comment if you'd like.

File: django/sierra/base/models.py [hunk 1]
```diff
@@ -1948,6 +1947,68 @@ def get_call_numbers(self):
...
+                    probably_shelved_by_title = True
+                    break
+
+        if probably_shelved_by_title:
+            bib = self.bibrecorditemrecordlink_set.all()[0].bib_record
+            title = bib.bibrecordproperty_set.all()[0].best_title
+            title = re.sub(r'\.*\s*$', r'', title)
+            call_number = '{} -- {}'.format(call_number, title)
+            cn_type = 'other'
+
```
Comment (damonkelley) [inline django/sierra/base/models.py]: It is hard to spot, but there is extra whitespace here.

File: django/sierra/base/models.py [hunk 2]
```diff
@@ -1948,6 +1947,68 @@ def get_call_numbers(self):
...
         return call_number_tuples
 
+    def get_shelving_call_number(self):
+        """
+        Returns one tuple that should represent the call number used
+        to shelve this item: (call_number, cn_type).
+        """
+        call_number, cn_type = (None, None)
+        item_cn_tuples = self.get_call_numbers()
+        bib_cn_tuples = []
```
Comment (damonkelley) [inline django/sierra/base/models.py]: Maybe consider moving this to the `except` block

**Comment by damonkelley** [comment]
It looks like `get_shelving_call_number` is mostly a cut and paste, so it is working code, and no need to mess with working code.

That said, I think there are some possible refactorings available to you if you think they are worth it. I think could extract a good portion of the `get_shelving_call_number` body into separate smaller _private_ methods. Especially, where you have large comment sections. Things like that usually lead me to think that the method is doing too much, and a new named function might be a good alternative. :smile:

File: django/sierra/base/models.py [hunk 3]
```diff
@@ -1948,6 +1947,68 @@ def get_call_numbers(self):
...
+        # periodicals) and the call number is just two letters (like
+        # MT), or if the word "periodical" appears in the call number,
+        # or if there's ANY c-tagged field on the item with "SHELVED BY
+        # TITLE." In these cases, we get the Bib title and append it to
+        # the end of the call number.
+
+        probably_shelved_by_title = (
+            self.itype.code_num == 5 and (not call_number
+                or re.search(r'^[A-Za-z]{,2}$', call_number))
+            or (call_number and
```
Comment (damonkelley) [inline django/sierra/base/models.py]: A little bit of extra whitespace at the end of this line too. FYI

File: django/sierra/base/models.py (hunk 1 above)
Comment (jthomale) [inline django/sierra/base/models.py]: I have a feeling my code is littered with this. I should probably have considered starting using a pep8 linter long ago. Thanks for pointing it out.

**Comment by jthomale** [comment]
@damonkelley I agree with you on the refactoring of ItemRecord.get_shelving_call_number, which I've taken a stab at in commit 29328f6. Is that a bit better?

**Comment by damonkelley** [comment]
Yeah, I think so

File: django/sierra/base/models.py [hunk 4]
```diff
@@ -1947,50 +1947,31 @@ def get_call_numbers(self):
...
+        return is_sudoc
 
+    def _item_is_probably_shelved_by_title(self, cn_string):
+        """
+        Takes a call number string and tries to determine whether this
+        item is probably shelved by title. If the ITYPE is 5 (bound
+        periodicals) and the call number is just one or two letters
+        (like M or MT), or if the word "periodical" appears in the call
+        number string, or if there's ANY c-tagged field on the item
+        with "SHELVED BY TITLE." Returns a boolean value.
```
Comment (somexpert) [inline django/sierra/base/models.py]: missing the conclusion of those group of `or`s. so what happens if any of those are true? is it that the item is probably shelved?

File: django/sierra/base/models.py [hunk 5]
```diff
@@ -1999,15 +1980,48 @@ def get_shelving_call_number(self):
...
 
-        if probably_shelved_by_title:
-            bib = self.bibrecorditemrecordlink_set.all()[0].bib_record
-            title = bib.bibrecordproperty_set.all()[0].best_title
-            title = re.sub(r'\.*\s*$', r'', title)
-            call_number = '{} -- {}'.format(call_number, title)
+    def _add_title_to_cn_string(self, cn_string):
+        """
+        Takes a call number string, appends the bib record title onto
+        onto the end, and returns it. Useful for generating the call
```
Comment (somexpert) [inline django/sierra/base/models.py]: `onto onto`

File: django/sierra/base/models.py [hunk 6]
```diff
@@ -1948,6 +1947,82 @@ def get_call_numbers(self):
...
+                    content_method='display_field_content'):
+                if cn.strip().upper() == 'SHELVED BY TITLE':
+                    probably_shelved_by_title = True
+                    break
+        return probably_shelved_by_title
+
+    def _add_title_to_cn_string(self, cn_string):
+        """
+        Takes a call number string, appends the bib record title onto
+        onto the end, and returns it. Useful for generating the call
```
Comment (somexpert) [inline django/sierra/base/models.py]: `onto onto`

**Comment by jthomale** [comment]
@somexpert Awesome--thanks for catching those. Fixed in b4a4f52.

**Comment by damonkelley** [comment]
:+1:

File: django/sierra/base/models.py [hunk 7]
```diff
@@ -1948,6 +1947,82 @@ def get_call_numbers(self):
...
+        """
+        Returns a tuple that should represent the call number used
+        to shelve this item: (cn_string, cn_type), where the cn_string
+        is usable for sorting a list of items in call number order
+        (after normalization).
+        """
+        item_cn_tuples = self.get_call_numbers()
+        try:
+            bib_cn_tuples = (self.bibrecorditemrecordlink_set.all()[0]
+                                .bib_record.get_call_numbers())
```
Comment (somexpert) [inline django/sierra/base/models.py]: Can we get the dots to align here?

**Comment by jthomale** [comment]
@somexpert Spacing/alignment problem fixed in ffbc0c5.

**Comment by somexpert** [comment]
:thumbsup:

**Pull request merged**